java/arcs/android/storage/database/AndroidSqliteDatabaseManager.kt

```diff
@@ -82,58 +84,59 @@ class AndroidSqliteDatabaseManager(
                 }
+            }.also {
+                dbCache[entry.name to entry.isPersistent] = it
+            }
         }
     }
 
     override suspend fun snapshotStatistics(): Map<DatabaseIdentifier, Snapshot> = mutex.withLock {
         dbCache.mapValues { it.value.snapshotStatistics() }
     }
 
-    override suspend fun resetAll() {
-        registry.fetchAll()
-            .map { getDatabase(it.name, it.isPersistent) as DatabaseImpl }
-            .forEach { it.reset() }
+    override suspend fun resetAll() = runOnAllDatabases { _, db ->
+        db.reset()
     }
 
-    override suspend fun removeExpiredEntities(): Job = coroutineScope {
-        launch {
-            registry.fetchAll()
-                .map { it.name to getDatabase(it.name, it.isPersistent) }
-                .forEach { (name, db) ->
-                    if (databaseSizeTooLarge(name)) {
-                        // If the database size is too large, we clear it entirely.
-                        db.removeAllEntities()
-                    } else {
-                        db.removeExpiredEntities()
-                    }
-                }
+    override suspend fun removeExpiredEntities() = runOnAllDatabases { name, db ->
+        if (databaseSizeTooLarge(name)) {
+            // If the database size is too large, we clear it entirely.
+            db.removeAllEntities()
+        } else {
+            db.removeExpiredEntities()
         }
     }
 
-    override suspend fun removeAllEntities() =
-        registry.fetchAll()
-            .map { getDatabase(it.name, it.isPersistent) }
-            .forEach { it.removeAllEntities() }
+    override suspend fun removeAllEntities() = runOnAllDatabases { _, db ->
+        db.removeAllEntities()
+    }
 
-    override suspend fun removeEntitiesCreatedBetween(startTimeMillis: Long, endTimeMillis: Long) =
-        registry.fetchAll()
-            .map { getDatabase(it.name, it.isPersistent) }
-            .forEach { it.removeEntitiesCreatedBetween(startTimeMillis, endTimeMillis) }
+    override suspend fun removeEntitiesCreatedBetween(
+        startTimeMillis: Long,
+        endTimeMillis: Long
+    ) = runOnAllDatabases { _, db ->
+        db.removeEntitiesCreatedBetween(startTimeMillis, endTimeMillis)
+    }
 
-    override suspend fun runGarbageCollection(): Job = coroutineScope {
-        launch {
+    override suspend fun runGarbageCollection() = runOnAllDatabases { _, db ->
+        db.runGarbageCollection()
+    }
+
+    /** Execute the provided block on each of the registered databases. */
+    private suspend fun runOnAllDatabases(block: suspend (name: String, db: Database) -> Unit) =
+        coroutineScope {
             registry.fetchAll()
-                .map { getDatabase(it.name, it.isPersistent) }
-                .forEach { it.runGarbageCollection() }
+                .map { it.name to getDatabase(it.name, it.isPersistent) }
+                .forEach {
+                    launch {
```

Comment:
Just for my understanding, can you explain me one thing, before we had (eg runGarbageCollection)
coroutineScope {
  launch {
     ... // do stuff
and it was returning a job. 

what we have now looks the same to me, coroutineScope+launch, just changing the loop structure.
how come this does not return a job now?

Reply:
In my original change here, the `coroutineScope` was removed completely.

In #5798, @jasonwyatt suggested a change along with this, which was to run the database jobs in parallel.

This is a perfect job for `coroutineScope`: To start a number of parallel jobs, and wait for them to complete.

The structure isn't quite the same here: there's not just one job now, since we will call `launch` once for each database.

And in any case, there's no need to return a job at all: when we've progressed past the `coroutineScope` block, all of the jobs launched of inside of it are guaranteed to be completed.

Reply:
In other words,
```
coroutineScope {
 launch { task() }
}
```

Doesn't give us anything, might as well just run `task()`

```
  coroutineScope {
    launch { task1() }
    launch { task2() }
    launch { task3() }
  }
```

Gives us a convenient way to run a number of tasks concurrently, and wait for their completion.

Reply:
I see, thanks for explaining. the change looks good.